src/NATS.Client.Core/NatsConnection.cs

```diff
@@ -594,6 +612,42 @@ private async void ReconnectLoop()
         }
     }
 
+    private async Task PublishEventsAsync()
+    {
+        try
+        {
+            while (!_disposedCancellationTokenSource.IsCancellationRequested)
+            {
+                var hasData = await _eventChannel.Reader.WaitToReadAsync(_disposedCancellationTokenSource.Token).ConfigureAwait(false);
```

Comment:
What's the benefit of serializing events over a channel and awaiting them, as opposed to just firing them off, wrapping them in a try/catch with a logger

I guess if there were a large amount of events we could be creating a lot of demand on the thread pool, but I don't think that is all too common?

Reply:
They're not serialized, just passed through a queue. One advantage of this intermediate step is that event delivery will be in-order, which cannot be guaranteed for fire-and-forget tasks that may run in random order (this could be mitigated by giving events a sequence number and/or timestamp, but recipients would then need to re-order events). It also allows for future pre-processing, e.g. to throttle error events, without blocking the connection.

Reply:
Probably a poor use of the word `serializing` - I meant what is the benefit of running one-at-a-time, awaiting each one.  In-order delivery certainly is an advantage, but this solution could suffer from head-of-line blocking

Reply:
Related note: One of the reasons for using a channel to deliver events is to be able to fire synchronously using `TryWrite` e.g. from subscriber's channel full `Action`.

Reply:
> In-order delivery certainly is an advantage, but this solution could suffer from head-of-line blocking

I suspect that slow-running handler code would have been much worse with the old code, as at least some were fired while locks were held. 

There are some things we could consider to perhaps improve the situation:
- assign a sequence number to events and fire them off asynchronously (but this needs task tracking)
- use a bounded channel with some discard strategy to avoid flooding consumers with events (but this may cause lost events for slow handlers)
- deduplication for queued events, i.e. only publish the last of a given type (not entirely trivial)
- throttling for error events (min-interval between reports, possibly with scaled delays; also not trivial)

However, I got the impression that we would move forward with the current solution and then iterate on it after discussing options in a follow-up thread (mostly for considering the benefits of whether to manage the delegates manually, so that we could support both sync and async event syntax, and same some allocations). Maybe we need to have this discussion before this PR is merged?

Reply:
I'm fine to move forward with the current solution, I think how the events are invoked can be changed in a non-breaking fashion.